### DISCUSSION TITLE
Ability to override operators such as the property accessor

### DISCUSSION THREAD

**Discussion by krashdifferent** [josdejong/mathjs#3054] [Ideas]
I am looking for a way to override operators in the expression evaluator, particularly the property accessor. Essentially I would like to apply an optional transform function to this operator but it is unclear to me if this is possible.

Specific use cases:
While I agree with the default behavior, there a couple example extensions I am looking to add that better fits our workflow. In this case users are writing a short script that gets run on a lot of object data entries.

1. Deeply nested objects are common in our data, however it is also common for an early value in the chain to be undefined or null. I would like to make attempting to access a property on undefined or null return undefined or null rather than throw an error. The expression `a.b.c.d` would return undefined if `a.b` is undefined.

2. It is also common for us to have arrays of objects. In this case I would like to have an accessor attempted on an array automatically map across all of the array elements. Obviously while explicitly calling map in the expression is possible it can get very verbose. If `a` is an array `a.b` would become equivalent to `map(a, f(x) = x.a)`

Thanks for the library!

**Comment by josdejong** [comment]
Thanks for your suggestions @krashdifferent. 

The property getters and setters of mathjs are currently hardcoded and cannot be customized. To make the expression parser secure, this getter and setter have special security rules built in (for example to prevent you from getting access to `someFn.constructor`). I think we should be really careful if we want to make the getter and setter customizable, we should not introduce new security issues.

Just thinking aloud: 

1. looks like you're looking for optional chaining (`a?.b?.c?.d`). We can consider building this in the expression parser.
2. can you maybe utilize the `get` function from lodash? Or write your own util function?
3. about (2), maybe you can write a little util function for that too? Something like:

    ```
    mapA(array) = map(array, f(item) = item.a)
    mapA(a)
    ```

**Comment by gwhitney** [comment]
This seems like more of an open-ended proposal. Moving to Discussions.
